pkg/crypto/certloader/watcher.go

```diff
@@ -0,0 +1,239 @@
+	}
+
+	w.Watch()
+	return w, nil
+}
+
+// EventualWatcher returns a channel where exactly one Watcher will be sent
+// once the given files are ready and loaded. This can be useful when the file
+// paths are well-known, but the files themselves don't exist yet. Note that
+// the requirement is that the file directories must exists.
+func EventualWatcher(log logrus.FieldLogger, caFiles []string, certFile, privkeyFile string) (<-chan *Watcher, error) {
+	r, err := NewFileReloader(caFiles, certFile, privkeyFile)
+	if err != nil {
+		return nil, err
+	}
+	fswatcher, err := newFsWatcher(caFiles, certFile, privkeyFile)
+	if err != nil {
+		return nil, err
+	}
+	w := &Watcher{
+		FileReloader: r,
+		log:          log,
+		fswatcher:    fswatcher,
+		stop:         make(chan struct{}),
+	}
+
+	res := make(chan *Watcher)
+	go func(res chan<- *Watcher) {
+		defer close(res)
+		// Attempt a reload without having received any fs notification in case
+		// all the files are already there. Note that the keypair and CA are
+		// reloaded separately as a "partial update" is still useful: If the
+		// FileReloader is "half-ready" (e.g. has loaded the keypair but failed
+		// to load the CA), we only need a successfully handled CA related fs
+		// notify event to become Ready (in other words, we don't need to
+		// receive a fs event for the keypair in that case to become ready).
+		_, keypairErr := w.ReloadKeypair()
+		_, caErr := w.ReloadCertificateAuthority()
+		ready := w.Watch()
+		if keypairErr == nil && caErr == nil {
+			log.Debug("TLS configuration ready")
+			res <- w
+			return
+		}
+		log.Debug("waiting on fsnotify update to be ready")
+		select {
+		case <-ready:
+			log.Debug("TLS configuration ready")
+			res <- w
+		case <-w.stop:
+		}
+	}(res)
+
+	return res, nil
+}
+
+// Watch initialize the files watcher and update goroutine. It returns a ready
+// channel that will be closed once an update made the underlying FileReloader
+// ready.
+func (w *Watcher) Watch() chan struct{} {
+	// build maps for the CA files and keypair files to help detecting what has
+	// changed in order to reload only the appropriate certificates.
+	keypairMap := make(map[string]struct{})
+	caMap := make(map[string]struct{})
+	if w.FileReloader.certFile != "" {
+		keypairMap[w.FileReloader.certFile] = struct{}{}
+	}
+	if w.FileReloader.privkeyFile != "" {
+		keypairMap[w.FileReloader.privkeyFile] = struct{}{}
+	}
+	for _, path := range w.FileReloader.caFiles {
+		caMap[path] = struct{}{}
+	}
+
+	// prepare the ready channel to be returned. We will close it exactly once.
+	var once sync.Once
+	ready := make(chan struct{})
+	go func() {
+		defer w.fswatcher.Close()
+		for {
+			select {
+			case event := <-w.fswatcher.Events:
+				path := event.Name
+				log := w.log.WithFields(logrus.Fields{
+					logfields.Path: path,
+					"operation":    event.Op,
+				})
+				log.Debug("Received fsnotify event")
+				_, keypairChanged := keypairMap[path]
+				_, caChanged := caMap[path]
```

Comment:
I had to give this a careful read to understand. Actually, it may be the same keypair fwiw but what `keypairChanged` actually means is that the path is now different, right?

So what's not clear to me is that if the keypair path changed, how would `ReloadKeypair()` ever succeed knowing that the corresponding path in the watcher is not updated? Please, correct me if I'm missing something.

Reply:
The paths are constants for the lifetime of the Watcher. 

`keypairMap` and `caMap` are simply map with paths as keys. `keypairChanged` and `caChanged` are flags indicating whether the received fsnotify event is about a keypair file (respectively CA file).